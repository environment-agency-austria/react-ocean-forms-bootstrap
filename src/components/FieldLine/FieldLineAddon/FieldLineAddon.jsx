/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { InputGroupAddon, InputGroupText } from 'reactstrap';
import { FormText } from 'react-ocean-forms';

/**
 * Displays a prefix / suffix addon for the FieldLine
 */
function FieldLineAddon({ plaintext, content, type }) {
  if (plaintext) return null;
  if (!content) return null;

  let child = null;
  if (typeof content === 'string') {
    child = (
      <InputGroupText>
        <FormText text={content} />
      </InputGroupText>
    );
  } else {
    child = content;
  }

  return (
    <InputGroupAddon addonType={type}>
      {child}
    </InputGroupAddon>
  );
}

FieldLineAddon.defaultProps = {
  content: null,
};

FieldLineAddon.propTypes = {
  type: PropTypes.string.isRequired,
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  plaintext: PropTypes.bool.isRequired,
};

export default FieldLineAddon;
