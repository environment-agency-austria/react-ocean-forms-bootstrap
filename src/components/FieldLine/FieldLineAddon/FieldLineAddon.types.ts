export type IFieldLineAddonType = 'prepend' | 'append';

/**
 * Props for the FieldLineAddon component
 */
export interface IFieldLineAddonProps {
  /**
   * Addon type
   */
  type: IFieldLineAddonType;
  /**
   * Form plaintext mode
   */
  plaintext: boolean;
  /**
   * Addon content / child
   */
  content?: React.ReactNode;
}
