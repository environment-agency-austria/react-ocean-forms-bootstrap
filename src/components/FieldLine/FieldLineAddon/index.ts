export * from './FieldLineAddon';
export * from './FieldLineAddon.types';
