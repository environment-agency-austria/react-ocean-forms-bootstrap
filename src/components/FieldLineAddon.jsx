/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { InputGroupAddon } from 'reactstrap';
import { fieldMetaShape } from 'react-ocean-forms';

/**
 * Displays a prefix / suffix addon for the FieldLine
 */
function FieldLineAddon({ meta, content, type }) {
  if (meta.plaintext) return null;
  if (!content) return null;

  let child = null;
  if (typeof content === 'string') {
    child = meta.stringFormatter(content);
  } else {
    child = content;
  }

  return (
    <InputGroupAddon addonType={type}>
      {child}
    </InputGroupAddon>
  );
}

FieldLineAddon.defaultProps = {
  content: null,
};

FieldLineAddon.propTypes = {
  type: PropTypes.string.isRequired,
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  meta: fieldMetaShape.isRequired,
};

export default FieldLineAddon;
