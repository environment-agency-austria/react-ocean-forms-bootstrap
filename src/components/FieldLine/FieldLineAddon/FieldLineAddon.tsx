/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';

import { FormText } from 'react-ocean-forms';
import { InputGroupAddon, InputGroupText } from 'reactstrap';

import { IFieldLineAddonProps } from './FieldLineAddon.types';

/**
 * Displays a prefix / suffix addon for the FieldLine
 */
export const FieldLineAddon: React.FC<IFieldLineAddonProps> = (props) => {
  const { plaintext, content, type } = props;

  if (plaintext || content === undefined) { return null; }

  let child = null;
  if (typeof content === 'string') {
    child = (
      <InputGroupText>
        <FormText text={content} />
      </InputGroupText>
    );
  } else {
    child = content;
  }

  return (
    <InputGroupAddon addonType={type}>
      {child}
    </InputGroupAddon>
  );
};
