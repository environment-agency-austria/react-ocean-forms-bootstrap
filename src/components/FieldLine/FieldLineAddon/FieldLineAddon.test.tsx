import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { FormText } from 'react-ocean-forms';

import { FieldLineAddon } from './FieldLineAddon';
import { IFieldLineAddonType } from './FieldLineAddon.types';

describe('<FieldLineAddon />', () => {
  const setup = (
    plaintext: boolean,
    content?: React.ReactNode,
    type: IFieldLineAddonType = 'prepend',
  ): ShallowWrapper => {
    return shallow((
      <FieldLineAddon
        plaintext={plaintext}
        content={content}
        type={type}
      />
    ));
  };

  const checkAddonRender = (wrapper: ShallowWrapper, type: IFieldLineAddonType): void => {
    it('should render an InputGroupAddon', () => {
      expect(wrapper.find('InputGroupAddon').exists()).toBeTruthy();
    });

    it('should pass the correct addon type', () => {
      expect(wrapper.find('InputGroupAddon').prop('addonType')).toBe(type);
    });
  };

  it('should render nothing if meta.plaintext is active', () => {
    const wrapper = setup(true);
    expect(wrapper.find('InputGroupAddon').exists()).toBeFalsy();
  });

  it('should render nothing if there is no content', () => {
    const wrapper = setup(false);
    expect(wrapper.find('InputGroupAddon').exists()).toBeFalsy();
  });

  describe('text content', () => {
    const mockContent = 'mock-string';
    const mockType = 'prepend';
    const wrapper = setup(false, mockContent, mockType);

    describe('render', () => {
      checkAddonRender(wrapper, mockType);

      it('should wrap the text content in a FormText', () => {
        const formText = wrapper.find(FormText);
        expect(formText).toHaveLength(1);
        expect(formText.prop('text')).toBe(mockContent);
      });

      it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('custom content', () => {
    const mockContent = <div id="mock-content" />;
    const mockType = 'append';
    const wrapper = setup(false, mockContent, mockType);

    describe('render', () => {
      checkAddonRender(wrapper, mockType);

      it('should wrap the text content in a FormText', () => {
        expect(wrapper.find(FormText)).toHaveLength(0);
      });

      it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
