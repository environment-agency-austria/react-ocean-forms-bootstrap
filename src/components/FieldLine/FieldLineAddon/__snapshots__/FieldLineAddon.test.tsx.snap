// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldLineAddon /> custom content render should render correctly 1`] = `
<InputGroupAddon
  addonType="append"
  tag="div"
>
  <div
    id="mock-content"
  />
</InputGroupAddon>
`;

exports[`<FieldLineAddon /> text content render should render correctly 1`] = `
<InputGroupAddon
  addonType="prepend"
  tag="div"
>
  <InputGroupText
    tag="span"
  >
    <FormText
      text="mock-string"
    />
  </InputGroupText>
</InputGroupAddon>
`;
